source/common/router/router.cc

```diff
@@ -841,9 +838,15 @@ Http::FilterDataStatus Filter::decodeData(Buffer::Instance& data, bool end_strea
     callbacks_->addDecodedData(many_copied_buffer.nextBuffer(), true);
   }
   if (streaming_shadows_ && !shadow_streams_.empty()) {
-    for (auto& [shadow_stream, ended] : shadow_streams_) {
-      shadow_stream->captureAndSendData(many_copied_buffer.nextBufferOwned(), end_stream);
-      ended = end_stream;
+    for (auto* shadow_stream : shadow_streams_) {
+      if (end_stream) {
+        shadow_stream->removeDestructorCallback();
+        shadow_stream->removeWatermarkCallbacks();
+      }
+      shadow_stream->sendData(many_copied_buffer.nextBuffer(), end_stream);
+    }
+    if (end_stream) {
+      shadow_streams_.clear();
```

Comment:
So having told you to do this I have another question I want to ask because I'm second guessing =P
if we haven't spooled all the data to a shadow stream and the client disconnects, do we want to cancel, or not?  I'm inclined to say it'd better to be consistent (send all the data read, which is what we do now) but I wanted to check
If we leave as-is let's at least comment here that clearing the shadow streams doesn't indicate they're done - they're self owning and will spool data to the upstream on their own, we just stop the assocation.

Also sanity check, if they're disassociated and the upstream stops reading, will they eventually time out?

Reply:
Not totally sure what you mean -- these are the scenarios I can think of:
1. We see `end_stream`, and then the client disconnects before the main upstream has properly sent a response. In this case, the shadow streams are already disassociated from the main stream, and they'll continue on their way, respecting the timeouts associated with the shadow cluster.
2. The client disconnects before sending `end_stream`. Here, we'll (eventually) destroy the router filter, which will cancel all the shadow streams.